Enhancements for the Authentication mechanisms documentation (#30658)
quarkusio/quarkus: docs/src/main/asciidoc/security-authentication-mechanisms-concept.adoc

include::_attributes.adoc[]
:categories: security,web

The Quarkus Security framework supports multiple authentication mechanisms, which you can use to secure your applications.
Some supported authentication mechanisms are built into Quarkus, and some require you to add an extension.
You can also combine authentication mechanisms.

Before you choose an authentication mechanism for securing your Quarkus applications, review the information provided.

.Authentication requirements and mechanisms
|===
^|Authentication Requirement ^| Authentication Mechanism ^| Notes

^|Username and password ^|xref:security-basic-authentication-concept.adoc[Basic], xref:security-authentication-mechanisms-concept.adoc#form-auth[Form] ^|

> ^|Bearer Access Token  ^| xref:xref:security-openid-connect-web-authentication.adoc[OIDC], xref:security-jwt.adoc[JWT], xref:security-oauth2.adoc[OAuth2]  ^|See the <<table>> table below.

Reviewer: This link should point to the bearer access token doc, currently named `security-openid-connect.adoc`

Reviewer: I see, double `xref` here

Author: Fixing

Reviewer: @MichalMaler The other thing is that `Notes` column seems redundant right now, it was my idea I know, but it does not look good yet while we only have a single note, as it creates a widest column, with all but only a single value set, my proposal is to collapse that note into the 2nd column `OIDC, JWT, OAuth2 (see <<table>> table below)`

Author: Yeah, I already tried that. The problem is that it doenst look well and asciidoc doesnt allowed me to use bullet point for that Table-note. Need to find a way how to do it. Working on it

Author: Nice, found a solution. Adding.

Author: @sberyozkin by that file `security-openid-connect.adoc`, did you meant `security-openid-connect-client.adoc` ?

Author: it was this one I guess `security-oidc-bearer-authentication-concept.adoc` Renamed and recomposed by Rolfe.

> ^|Single Sign-On (SSO) ^|xref:xref:security-openid-connect-web-authentication.adoc[OIDC], xref:security-authentication-mechanisms-concept.adoc#form-auth[Form]   ^|

Reviewer: ```suggestion
^|Single Sign-On (SSO) ^|xref:security-openid-connect-web-authentication.adoc[OIDC], xref:security-authentication-mechanisms-concept.adoc#form-auth[Form]   ^|
```

Author: nice catch

include::_attributes.adoc[]
:categories: security,web

The Quarkus Security framework supports multiple authentication mechanisms, which you can use to secure your applications.
Some supported authentication mechanisms are built into Quarkus, and some require you to add an extension.
You can also combine authentication mechanisms.

Before you choose an authentication mechanism for securing your Quarkus applications, review the information provided.

.Authentication requirements and mechanisms
[source, adoc]
|===
^|Authentication Requirement ^| Authentication Mechanism

^|Username and password ^|xref:security-basic-authentication-concept.adoc[Basic], xref:security-authentication-mechanisms-concept.adoc#form-auth[Form]

^|Bearer Access Token  ^a| xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer], xref:security-jwt.adoc[JWT], xref:security-oauth2.adoc[OAuth2]

> * See the <<table>> table below.

Reviewer 2: In the renderend output, this is the only table entry that has also has an entry in the Notes column. When this column is blank for all the other items in the table, it looks a bit funny and also might pose an issue for Accessibility.  Would it be possible to remove the Notes column and move this one item underneath the table (perhaps as a callout)

Author: Hello Sheila @sheilamjones . Rendered output was just for content moving. It is very obsolete right now. You can have a new rendered view when you go: Files changed -> then in the right upper corner -> ... -> View file.
Please, always use the last one.
Adding it here:
https://github.com/quarkusio/quarkus/blob/d0f419683041593c0a792311fdf37149083833a3/docs/src/main/asciidoc/security-authentication-mechanisms-concept.adoc

Author: @sberyozkin @sheilamjones Hello. Please, where exactly in the latest build, which is the only relative built when providing a review, do you see that extra Notes column? :(
![image](https://user-images.githubusercontent.com/48474054/215998167-09190b4b-0d80-47ea-bf62-255d3884f32c.png)

^|Single Sign-On (SSO) ^|xref:security-oidc-code-flow-authentication-concept.adoc[OIDC Web], xref:security-authentication-mechanisms-concept.adoc#form-auth[Form]

^|Client Certificate  ^|xref:security-authentication-mechanisms-concept.adoc#mutual-tls[Mutual TLS (MTLS)]

^|WebAuthn ^|xref:security-webauthn-concept.adoc[WebAuthn]

> ^|Kerberos ticket    ^|link:https://quarkiverse.github.io/quarkiverse-docs/quarkus-kerberos/dev/index.html[Quarkus - Kerberos]

Reviewer: ```suggestion
^|Kerberos ticket    ^|link:https://quarkiverse.github.io/quarkiverse-docs/quarkus-kerberos/dev/index.html[Kerberos]
```

Reviewer: These are all Quarkus extensions, so propose to remove `Quarkus -`. This is the last change proposal from me, thanks @MichalMaler

|===

== Built-in authentication support

Quarkus Security provides the following built-in authentication mechanisms:

* xref:security-basic-authentication-concept.adoc[Basic authentication]
* <<form-auth>>

> * <<mutual-tls>>

Reviewer 2: In the rendered preview, there are brackets "[ ]" appearing around the this "mutual-tls" link. Just want to double-check that is rendering correctly from your side.

Reviewer: I saw it as well

Author: It renders correctly in the latest version of a preview.

[[oidc-jwt-oauth2-comparison]]

== Choose the authentication mechanism

> Quarkus provides the following authentication mechanisms to secure your Quarkus applications:

Reviewer 3: This IMO makes it look like it is an exhaustive list (and that Quarkus doesn't support anything beyond these 3), which isn't true. It's just that this section attempts to explain the differences between these 3 mechanisms in particular, but there are more (for example, just a hardcoded list of users and passwords).

Author: Hm, ok. Will put it back. Thank you

== Choosing between OpenID Connect, SmallRye JWT, and OAuth2 authentication mechanisms

> Use the following information to select the proper authentication mechanism for your applications.

Reviewer: ```suggestion
Use the following information to select the proper token authentication mechanism for your applications.
```

== Choosing between OpenID Connect, SmallRye JWT, and OAuth2 authentication mechanisms

> Use the following information to select the proper token authentication mechanism for your applications.

Reviewer 2: Style suggestion: I think the use of "proper" here could be a bit ambiguous as in, would a user know which is proper mechanism for their application and which is now? I think "appropriate" might be better.
Nitpick: For maximum clarity, I think I would keep the previous wording for the end of the sentence, that is,  "Use the following information to select the appropriate token authentication mechanism to secure your Quarkus applications."

Author: OK, will apply.

* If your Bearer tokens must be verified, use `quarkus-oidc`, `quarkus-smallrye-jwt`, or `quarkus-elytron-security-oauth2`.

> * If your Bearer tokens are in a JWT format, you can use either of the three extensions.

Reviewer 4: ```suggestion
* If your Bearer tokens are in a JWT format, you can use any of the extensions listed above.
```

Both `quarkus-oidc` and `quarkus-smallrye-jwt` support refreshing the JsonWebKey (JWK) set when the OpenID Connect provider rotates the keys.
Therefore, if remote token introspection must be avoided or is unsupported by the providers, use `quarkus-oidc` or `quarkus-smallrye-jwt` for verifying JWT tokens.

* If you need to introspect the JWT tokens remotely, you can use either `quarkus-oidc` or `quarkus-elytron-security-oauth2` because they support the verification of the opaque or binary tokens by using remote introspection.
`quarkus-smallrye-jwt` does not support the remote introspection of both opaque or JWT tokens but instead relies on the locally available keys that are usually retrieved from the OpenID Connect provider.

* `quarkus-oidc` and `quarkus-smallrye-jwt` support the JWT and opaque tokens injection into the endpoint code.
Injected JWT tokens provide more information about the user.

> All extensions can have the tokens injected as `Principal`.

Reviewer 4: Will the target audience know what this means?

Author: I guess they will :) Thank you for the rev!

[NOTE]
====
Architectural considerations drive your decision to use opaque or JWT token format.
Opaque tokens tend to be much shorter than JWT tokens but need most of the token-associated state to be maintained in the provider database.
Opaque tokens are effectively database pointers.
JWT tokens are significantly longer than opaque tokens. Still, the providers effectively delegate most of the token-associated state to the client by storing it as the token claims and either signing or encrypting them.
====

> .Credential types and authentication mechanisms

Reviewer: This table is a higher level table which should not be placed here in the section comparing OIDC/JWT/OAuth2, 
I'd consider placing at the very top of the doc just above Built in mechanisms. Or may be in the main security concept

Author: Moving it to the top for now. Will fix the rest and let you know. Thank you!

|===
^|Credential type ^| Authentication ^| Notes

> ^|Username and password ^|xref:security-basic-authentication-concept.adoc[Basic ], xref:security-basic-authentication-concept.adoc#form-auth[Form], xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer] ^|

Reviewer: ```suggestion
^|Username and password ^|xref:security-basic-authentication-concept.adoc[Basic], xref:security-basic-authentication-concept.adoc#form-auth[Form] ^|
```

^|Bearer Access Token  ^| xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer], xref:security-jwt.adoc[JWT], xref:security-oauth2.adoc[OAuth2]  ^|See the <<table>> table below.

> ^|Single Sign-On (SSO) ^|xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer], xref:security-basic-authentication-concept.adoc#form-auth[Form]   ^|

Reviewer: It is a built-in authentication where the form auth is described

Author: So, do you want me to remove the SSO row completly, or replace SSO with the Built-in Aut?
@sberyozkin

Reviewer: @MichalMaler No, I meant that the form link is wrong, points to the basic auth

Author: Fixing. My bad.

> ^|Single Sign-On (SSO) ^|xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer], xref:security-basic-authentication-concept.adoc#form-auth[Form]   ^|

Reviewer: We should probably rename the `Credential Type` header to `Authentication Requirement` (and the next one to `Authentication Mechanism`) as SSO is not really a type and for ex Keycloak can use many forms of authentication in the SSO flows (I know I proposed to list it here :-) ).
Also the OIDC link should link to the oidc web authentication concepr

Author: OK, renaming. :)

Reviewer 3: I thought the goal of this table was to purely list credential types (password, certificate, token), and that "authentication requirements" are rather covered in the existing table (by that I mean token verification, single-sign on, multi-tenancy,...), no?

^|Client Certificate  ^|xref:security-authentication-mechanisms-concept.adoc#mutual-tls-authentication[Mutual TLS (MTLS)]   ^|

^|WebAuthn ^|xref:security-webauthn-concept.adoc[WebAuthn] ^|

^|Kerberos ticket    ^|link:https://quarkiverse.github.io/quarkiverse-docs/quarkus-kerberos/dev/index.html[Quarkus - Kerberos]   ^|
|===

[[table]]

> .Quarkus authentication extensions summary

Reviewer: ```suggestion
.Token authentication mechanism features
```

[NOTE]
====

> Architectural considerations drive your decision to use opaque or JWT token format.

Reviewer 4: ```suggestion
To decide whether to use opaque or JWT token format, consider your hardware architecture.
```

Opaque tokens tend to be much shorter than JWT tokens but need most of the token-associated state to be maintained in the provider database.
Opaque tokens are effectively database pointers.

> JWT tokens are significantly longer than opaque tokens. Still, the providers effectively delegate most of the token-associated state to the client by storing it as the token claims and either signing or encrypting them.

Reviewer 4: ```suggestion
JWT tokens are significantly longer than opaque tokens. However, the providers delegate most of the token-associated state to the client by storing it as the token claims and either signing or encrypting them.
```

====

[[table]]
.Token authentication mechanism comparison
|===
^|Feature required 3+^| Authentication mechanism

^| ^s|`quarkus-oidc` ^s|`quarkus-smallrye-jwt` ^s| `quarkus-elytron-security-oauth2`

s|Bearer JWT verification ^|Local verification or introspection  ^|Local verification  ^|Introspection

s|Bearer opaque token verification ^|Introspection   ^|No  ^|Introspection
s|Refreshing `JsonWebKey` set for verifying JWT tokens  ^|Yes   ^|Yes  ^|No
s|Represent token as `Principal`  ^|Yes   ^|Yes  ^|Yes

> s|Inject JWT as MP JSON Web Token (JWT)  ^|Yes   ^|Yes  ^|No

Reviewer 2: Style/Capitalization: 
- Use sentence case - "Inject JWT as MP JSON web token 
- Is it necessary to have (JWT) written in brackets in this instance given that it is already used in a few previous instances in the table. Would it be possible to instead place it after the very first occurrence in the topic?

Reviewer: `JSON Web Token` is how it is used on the web/everywhere, it is OK IMHO

Author: Will do. Thx for the catch

> s|Authorization Code Flow  ^| Yes  ^|No  ^|No

Reviewer 2: Capitalization: propose to change to "Authorization code flow"

s|Multi-tenancy ^| Yes  ^|No  ^|No
s|User information support  ^| Yes  ^|No  ^|No

> s|PEM key format support  ^|No   ^|Yes  ^|No

Reviewer 2: Should "PEM" be written out in full here for this first instance?

Author: OK, will add.
